### ISSUE TITLE
API Development Part 1 (OpenID Connect Infrastructure)

### ISSUE THREAD

**Issue by KentShikama** [diaspora/diaspora#6182]
If you are planning to join in right now, please see the [part 2 issue for the API](https://github.com/diaspora/diaspora/issues/6289)

The original [PR](https://github.com/diaspora/diaspora/pull/6095) is now getting kind of long and hard to follow if you are just starting. Hence, I'm creating this issue to inform everyone where we are at in terms of the API development and to breakdown what needs to be done next so it will be easier for anyone interested to hop in.
## [Resource Owner Password Credentials Grant](https://tools.ietf.org/html/rfc6749#section-4.3)

Note this flow is NO LONGER supported (Revert ee7f188ceff to re-enable). In case you still want to test it out after you revert the referenced commit, here is the complete flow using cURL. See tests for more "documentation".

1) Register your client

`curl http://localhost:3000/api/openid_connect/clients -d "redirect_uris[]=http://localhost:3000/&client_name=<client_name_here>"`

Note that the redirect uri isn't actually used in this flow. It should return the following:

```
{
 "o_auth_application":
  {
    "id":3,
    "client_id":<your_client_id>",
    "client_secret":"<your_client_secret>",
    "name":null,
    "redirect_uris":"[\"http://localhost:3000/\"]",
    "created_at":"2015-07-07T17:30:08.248Z",
    "updated_at":"2015-07-07T17:30:08.259Z"
  }
}
```

2) Retrieve an access token from the token endpoint 

`curl -d "grant_type=password&username=<your_username>&password=<your_password>&client_id=<your_client_id_from_above>&client_secret=<your_client_secret_from_above>" http://localhost:3000/api/openid_connect/access_tokens`

It should return the following:

```
{
  "access_token":"<your_access_token>",
  "token_type":"bearer",
  "expires_in":86399
}
```

3) Retrieve user info using your access token

Via Authorization Request Header Field
`curl -H "Authorization: Bearer <your_access_token>" http://localhost:3000/api/v0/user/`

Via Form-encoded Body Parameter
`curl -d "access_token=<your_access_token>" http://localhost:3000/api/v0/user/`

Via URI Query Parameter
`curl http://localhost:3000/api/v0/user/?access_token=<your_access_token>`

It should return the following:

```
{
  "name":"<your_username>@localhost:3000",
  "email":"<your_email>",
  "language":"<your_lang>",
  "username":"<your_username>"
}
```

<h3>Remaining tasks</h3>

- [x] Add support for [refresh tokens](https://tools.ietf.org/html/rfc6749#section-1.5) (See 8e55da3 and dc925f4 by @AugierLe42e )
- [x] ~~Add protection/tests "against brute force attacks (e.g., using rate-limitation or generating alerts)" as [specified](https://tools.ietf.org/html/rfc6749#section-4.3.2)~~ (No longer needed as this flow is disabled)
## [Dynamic client registeration](https://openid.net/specs/openid-connect-registration-1_0.html)

We are almost done here. See commit [8d2b5cd].
## [Implicit Flow](http://openid.net/specs/openid-connect-core-1_0.html#ImplicitFlowAuth)

Here is the current flow using cURL. Note I will assume your pod URL is `http://localhost:3000/`. @AugierLe42e has recently started [Griaspora](https://github.com/AugierLe42e/Griaspora), a client that plans to use this flow.

1) Register your client

`curl http://localhost:3000/api/openid_connect/clients -d "redirect_uris[]=http://localhost:3000/&client_name=<your_client_name>"`

It should return the following:

```
{
 "o_auth_application":
  {
    "id":3,
    "client_id":<your_client_id>",
    "client_secret":"<your_client_secret>",
    "name":"<your_client_name>",
    "redirect_uris":"[\"http://localhost\"]",
    "created_at":"2015-07-07T17:30:08.248Z",
    "updated_at":"2015-07-07T17:30:08.259Z"
  }
}
```

2) Run `bundle exec rake db:seed` to load the scopes

3) Send an Authentication Request to start off the Implicit flow

`curl -L  http://localhost:3000/api/openid_connect/authorizations/new?client_id=<your_client_id>\&redirect_uri=http://localhost:3000/\&response_type=id_token%20token\&scope=openid%20read\&nonce=hi\&state=hi --cookie-jar cookie`

It should return HTML code with the following snippet. The authenticity_token value is needed in the next step to pass the CSRF filter.

```
...

<form class="block-form" id="new_user" action="/users/sign_in" accept-charset="UTF-8" method="post"><input name="utf8" type="hidden" value="&#x2713;" /><input type="hidden" name="authenticity_token" value="7aKgpujMvPwmjBrzcGFIIRcsCzg+JY9u3MqaqS8D4mvBxfL0klexwgUa63WhSrMtq7DvKYARTexi8rGe3TFzmw==" />  <fieldset>

...more HTML
```

4) Login to Diaspora (with cURL)

`curl -L -H "X-CSRF-TOKEN: <your_authenticity_token_value_from_above>" -d "user[username]=<your_username>&user[password]=<your_password>&user[remember_me]=1" http://localhost:3000/users/sign_in --cookie-jar cookie --cookie cookie`

It should return HTML code with the following snippet.

```
...

<form class="approve" action="/openid_connect/authorizations" accept-charset="UTF-8" method="post"><input name="utf8" type="hidden" value="&#x2713;" /><input type="hidden" name="authenticity_token" value="a2lqmSwa4Bhov8SbKCVvt/PYUTDX0y/Efnfq1vj/1ei6J6awPrqi944gMcupSxZriLAZrgPZAFSAEQ0KokwO5Q==" />
      <input type="hidden" name="client_id" id="client_id" value="4bb87c09fd99949d74a0ad0af20278bf" />

...more HTML
```

5a) If the user has not yet approved the application, approve (or deny) the request

To approve

`curl -H "X-CSRF-TOKEN: <your_authenticity_token_value_from_above>" -d "approve=true" http://localhost:3000/api/openid_connect/authorizations --cookie-jar cookie --cookie cookie`

To deny

`curl -H "X-CSRF-TOKEN: <your_authenticity_token_value_from_above>" -d "approve=false" http://localhost:3000/api/openid_connect/authorizations --cookie-jar cookie --cookie cookie`

If you approve, it should return a ID token and access token in a fragment as follows. The ID token should contain two periods.

```
<html><body>You are being <a href="http://localhost:3000/#access_token=<your_access_token_value>&amp;expires_in=86399&amp;id_token=<your_id_token_value>&amp;token_type=bearer">redirected</a>.</body></html>
```

5b) If the user has already approved the application once

It should return a ID token and access token in a fragment as follows:

```
<html><body>You are being <a href="http://localhost:3000/#access_token=<your_access_token_value>&amp;expires_in=86399&amp;id_token=<your_id_token_value>&amp;token_type=bearer">redirected</a>.</body></html>
```

6) Verify the ID token

See `spec/controllers/openid_connect/id_tokens_controller_spec.rb` and `spec/controllers/openid_connect/authorizations_controller_spec.rb#create`.

7) Verify the Access Token using the ID token's at_hash value

See `spec/controllers/api/openid_connect/authorizations_controller_spec.rb#create` (specifically the test "it should return a valid access token in a fragment").

8) Retrieve user info using your access token 

Same as password flow step 3.

<h3>Remaining tasks</h3>

- [x] Add support for `token` response type (and `id_token token`)
- [x] Implement a real ID token. This involves creating the model, replacing the current random hex stub with the token generation, and adding tests.
- [x] Implement "read" and "write" scopes. 
- [x] Create/verify a script that validates ID tokens (See `id_tokens_controller_spec.rb`)
## [Authorization Code Flow](http://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth)

1) Run `bundle exec rake db:seed` to load the scopes

2) Register your client

`curl http://localhost:3000/api/openid_connect/clients -d "redirect_uris[]=http://localhost:3000/&client_name=<your_client_name>"`

It should return the following:

```
  {
    "id":3,
    "client_id":<your_client_id>",
    "client_secret":"<your_client_secret>",
    "name":"<your_client_name>",
    "redirect_uris":"[\"http://localhost:3000/\"]",
    "created_at":"2015-07-07T17:30:08.248Z",
    "updated_at":"2015-07-07T17:30:08.259Z"
  }
```

3) Send an Authentication Request to start off the Authorization Code flow

`curl -L  http://localhost:3000/api/openid_connect/authorizations/new?client_id=<your_client_id>\&redirect_uri=http://localhost:3000/\&response_type=code\&scope=openid%20read%20write\&nonce=hi\&state=hi --cookie-jar cookie`

It should return HTML code with the following snippet. The authenticity_token value is needed in the next step to pass the CSRF filter.

```
...

<form class="block-form" id="new_user" action="/users/sign_in" accept-charset="UTF-8" method="post"><input name="utf8" type="hidden" value="&#x2713;" /><input type="hidden" name="authenticity_token" value="7aKgpujMvPwmjBrzcGFIIRcsCzg+JY9u3MqaqS8D4mvBxfL0klexwgUa63WhSrMtq7DvKYARTexi8rGe3TFzmw==" />  <fieldset>

...more HTML
```

4) Login to Diaspora (with cURL)

`curl -L -H "X-CSRF-TOKEN: <your_authenticity_token_value_from_above>" -d "user[username]=<your_username>&user[password]=<your_password>&user[remember_me]=1" http://localhost:3000/users/sign_in --cookie-jar cookie --cookie cookie`

It should return HTML code with the following snippet.

```
...

<form class="approve" action="/openid_connect/authorizations" accept-charset="UTF-8" method="post"><input name="utf8" type="hidden" value="&#x2713;" /><input type="hidden" name="authenticity_token" value="a2lqmSwa4Bhov8SbKCVvt/PYUTDX0y/Efnfq1vj/1ei6J6awPrqi944gMcupSxZriLAZrgPZAFSAEQ0KokwO5Q==" />
      <input type="hidden" name="client_id" id="client_id" value="4bb87c09fd99949d74a0ad0af20278bf" />

...more HTML
```

5a) If the user has not yet approved the application, approve (or deny) the request

To approve

`curl -H "X-CSRF-TOKEN: <your_authenticity_token_value_from_above>" -d "approve=true" http://localhost:3000/api/openid_connect/authorizations --cookie-jar cookie --cookie cookie`

To deny

`curl -H "X-CSRF-TOKEN: <your_authenticity_token_value_from_above>" -d "approve=false" http://localhost:3000/api/openid_connect/authorizations --cookie-jar cookie --cookie cookie`

If you approve, it should return your auth code as a parameter as follows.

```
<html><body>You are being <a href="http://localhost:3000/?code=<your_auth_code_value>">redirected</a>.</body></html>
```

5b) If the user has already approved the application once

It should return an auth code as follows:

```
<html><body>You are being <a href="http://localhost:3000/?code=<your_auth_code_value>">redirected</a>.</body></html>
```

6) Send a token request to the token endpoint

`curl -d "grant_type=authorization_code&client_id=<your_client_id>&client_secret=<your_client_secret>&redirect_uri=http://localhost:3000/&code=<your_auth_code_value>" http://localhost:3000/api/openid_connect/access_tokens --cookie-jar cookie --cookie cookie`

It should return an ID token and access token as follows:

```
{
    "access_token":"<access_token_value>",
    "token_type":"bearer",
    "expires_in":86399,
    "id_token":"<id_token_value>"
}
```

7) Verify the ID token

See `spec/controllers/openid_connect/id_tokens_controller_spec.rb` and `spec/controllers/openid_connect/authorizations_controller_spec.rb#create`.

8) Verify the Access Token using the ID token's at_hash value

See `spec/controllers/api/openid_connect/authorizations_controller_spec.rb#create` (specifically the test "it should return a valid access token in a fragment").

9) Access protected resources using access token 

Same as password flow step 3.

<h3>Remaining tasks</h3>

- [x] Add support for `code` response type
- [x] Add necessary code to the token endpoint
## [Hybrid Flow](http://openid.net/specs/openid-connect-core-1_0.html#HybridFlowAuth)

See remaining tasks at the bottom of this post.
## [OpenID Connect Discovery](https://openid.net/specs/openid-connect-discovery-1_0.html)

Here we need to further adjust the discovery controller.

<h3>Remaining tasks</h3>

- [x] Add route, controller, and tests.
## Remaining Tasks for Initial API Release
- [x] Fix Pronto errors
- [x] Create views where user can reject/un-reject clients that have once been approved
- [x] Add support for [Prompt Parameter](http://openid.net/specs/openid-connect-core-1_0.html#ServerMTI)
- [x] Add support for [Display Parameter](http://openid.net/specs/openid-connect-core-1_0.html#ServerMTI)
- [x] Add support for [max_age Parameter](http://openid.net/specs/openid-connect-core-1_0.html#ServerMTI)
- [x] Use real 'sub' value/support for PPID
- [x] Implement request objects, including [require and require_uri](http://openid.net/specs/openid-connect-core-1_0.html#JWTRequests)
- [x] Add support for the [claims parameter](http://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter)
- [x] Add support for remaining [client attributes/metadata](https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata).
## Remaining Tasks for Future API Release

Or at least I will finish all the required API routes before coming back to these.

All of the below are optional based on the [OpenID Connect Specification](http://openid.net/specs/openid-connect-core-1_0.html)
- [ ] Background task for removing ID tokens
- [ ] Add support for `code token`, `code id_token` (Hybrid flows)
- [ ] Implement "extended" profile scope. The basic profile will just be the required "openid" scope. See [loomio discussion](https://www.loomio.org/d/ucvHeaOi/openid-connect-which-claims-to-communicate).
- [ ] Implement [signed request objects](http://openid.net/specs/openid-connect-core-1_0.html#JWTRequests)
- [ ] [Make user info access logs available to user](http://openid.net/specs/openid-connect-core-1_0.html#AccessMonitoring)
- [ ] Add mobile support for user end consent form
- [ ] Properly implement user hint
- [ ] Properly support ui_locales
- [ ] Support key rotation
- [ ] Add support for [`display=popup`](http://openid.net/specs/openid-connect-core-1_0.html#ServerMTI)
- [ ] Discuss whether we should allow Devise authorization in API endpoints
- [ ] [Polish error pages](https://github.com/diaspora/diaspora/pull/6095#issuecomment-155864757)
- [ ] Fix fallback for when logo_uri is pointing to a broken URL to not use the default avatar but the default application logo.
- [ ] Add rate-limiting for dynamic client registration
##

**Comment by AugierLe42e** [comment]
Thank you for taking time to do this ;)

**Comment by AugierLe42e** [comment]
> Implement a real ID token. This involves creating the model, replacing the current random hex stub with the token generation, and adding tests.

By what do you think we may replace the current random hex sting ?

**Comment by theworldbright** [comment]
> By what do you think we may replace the current random hex sting?

Do you mean when? Or how?  If you mean when, I was planning to implement this next. I'l be going to sleep now, so if you want to go ahead and implement it, feel free to do so. Or just inform me of what you plan to work on, and I'll start on something that doesn't conflict tomorrow. If you mean how, it might just be faster to read the spec and Nov's sample provider code than me explaining here (unless you have a specific question aspect about the ID token in mind?).

**Comment by AugierLe42e** [comment]
Now, I meant what is the problem with the current random hex? Why do you want to replace it, and by what. This is how Nov's sample works. I took this solution from it.

**Comment by theworldbright** [comment]
Hmm I believe you are talking about the access token? I'm talking about the ID token.

**Comment by AugierLe42e** [comment]
Ah! Ok, my bad, I misunderstood :)

**Comment by AugierLe42e** [comment]
[c13a923](https://github.com/diaspora/diaspora/commit/c13a923b2c086c1c5847227e5ac1ec751263f74a] adds support for authorizations by default. When an authorization doesn't already exists it is created by default on password flow.

**Comment by goobertron** [comment]
Just quickly: thank you both so much for all your work on this. Wishing you success!

**Comment by AugierLe42e** [comment]
Thank you very much @goobertron ;)

**Comment by theworldbright** [comment]
Thanks for the support @goobertron .

I added in the real ID token with the revised Authorization models. See Implicit flow 5b) above.

**Comment by theworldbright** [comment]
I added in support for access tokens in the implicit flow. See implicit flow steps 7) and 8) above. I will now work on adding support for request objects.

**Comment by theworldbright** [comment]
Actually since http://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter

"Support for the claims parameter is OPTIONAL. Should an OP not support this parameter and an RP uses it, the OP SHOULD return a set of Claims to the RP that it believes would be useful to the RP and the End-User using whatever heuristics it believes are appropriate."

I think we can just leave this be for now? @AugierLe42e I'm going to work on scopes.

**Comment by theworldbright** [comment]
Pardon me. Claims and requests objects are two different things. I will work on scopes first and then I will work on request objects. I will ignore claims until later.

**Comment by theworldbright** [comment]
Hmm never mind. Pardon me for all my changes. Claims are a large part of request objects because most other attributes that are in the request object itself are duplicated in the request, most notably, the client id and the scope. So I suppose I will be working with both after I finish with scopes.

**Comment by AugierLe42e** [comment]
Don't know. I haven't explored this part of the spec.

**Comment by AugierLe42e** [comment]
@theworldbright: refresh tokens implemented with https://github.com/AugierLe42e/diaspora/commit/8e55da39d4e87e64a26aa3925a27b17ec05d0b03 and https://github.com/AugierLe42e/diaspora/commit/dc925f40cd0176b6103a2e02c110b2ff98e686dc. You can check this.

**Comment by theworldbright** [comment]
Nice :+1:  I'll take a look soon.

**Comment by theworldbright** [comment]
I've updated this post with the latest changes.

**Comment by jaywink** [comment]
Awesome work guys <3

**Comment by AugierLe42e** [comment]
@theworldbright : I have reset your last changes and just neutralized the code ;)

**Comment by jaywink** [comment]
@AugierLe42e I doubt commented out code will be accepted to be merged in... please, just don't do it :)

**Comment by AugierLe42e** [comment]
Nope, it's just temporary. We'll see if we want to reactivate it when we start to work on OAuth for the API.

**Comment by theworldbright** [comment]
We do want to merge in this PR in before we start the API routes though right? @AugierLe42e @jhass 

I'm fine with keeping it commented out for now, although since we're on our last stretch I might just continue to work off my current branch (without the neutralization commit) and just push --force when I'm done (unless @AugierLe42e are you planning have a coding session this weekend?)

Also I've updated the original post (at the bottom) with things that I believe we should do before we start the API routes and things that we can leave until after.

**Comment by jhass** [comment]
Yes, I wouldn't mind having this PR in as early as possible. Then maybe a second one to setup a proper authorization infrastructure and then we can design and spread the API routes across many.

**Comment by AugierLe42e** [comment]
Ok, so I'll keep the code in a separate branch on my repo, you can force push @theworldbright.
I'll be free for coding this week-end, so we can keep in touch to finish this early. Maybe this week-end?
Then, I'll work on Griaspora and make an android mock to test it.
We can also plan to develop a "Connect with diaspora*" widget?

**Comment by theworldbright** [comment]
@AugierLe42e Yeah I'll definitely continue working this weekend. I just did the force push. Perhaps you can choose some task that is under the "Remaining Tasks for Initial API Release" that I haven't marked as "will do over the weekend"? (If you are interested in doing one that I have marked, just message me). Lets try to get the PR merged by the end of the weekend.

Yeah a "Connect with diaspora*" widget sounds great. Perhaps even create some branding guidelines for it. Maybe @pablocubico would be interested? I suppose we should create a new issue for this though. I'm fairly interested in having an android app that I can handle posts and chats through.

@jhass Good to know that you don't mind having the PR in soon. I'm really excited at the prospect of having the PR merged :p I'll create a "API Part 2" issue as a design proposal for how the API routes may look like soon.

**Comment by pablocubico** [comment]
@theworldbright Of course, I've been doing some typographic experiments for my pod (and some styling), I'll be happy to try and design a widget for that. I think that feature is key.

As for guidelines, I feel like I can only give recommendations, as there is no proper "design signoff" flow in the diaspora community, and design is still operating in an "I like / I don't like" basis, which is natural, but I like to support my design on data, not personal taste.

Anyone wanting to sketch a "Connect with diaspora" widget at least on pen and paper? Just to check some possible layouts.

**Comment by AugierLe42e** [comment]
@theworldbright: I can do the front-end part (forms, user consent, etc.) if you are ok with it.
I'll also start the "connect with" button during the week-end. I think it will be quick. I'll make some basic button and leave the styling to @pablocubico if you are ok with it.

**Comment by theworldbright** [comment]
@pablocubico Glad you're up for designing the widget :D

@AugierLe42e Yeah I'll leave the front-end part to you: styling the user consent form and then creating another page where you can manage OAuthApplications (basically the github equivalent of https://github.com/settings/applications ).

**Comment by theworldbright** [comment]
I'm going to add "[Make user info access logs available to end user](http://openid.net/specs/openid-connect-core-1_0.html#AccessMonitoring)" to "Remaining Tasks for Initial API Release".

**Comment by AugierLe42e** [comment]
@theworldbright, you can check *Create views where user can reject/un-reject clients that have once been approved* BTW ;)

**Comment by theworldbright** [comment]
Yeah I feel like we might have to adjust it based on the whether we are going to add more scopes.

**Comment by AugierLe42e** [comment]
From a design point of view, you mean? Otherwise, you would just need to add a new item in [the *scopes* section](https://github.com/diaspora/diaspora/pull/6095/files#diff-830f64755be0b6109a343e9889697243R1487) in the translations files.

**Comment by theworldbright** [comment]
Maybe some indented lists (if we support scopes that include other scopes) or something else that I haven't thought of that may come up in this discussion: https://github.com/diaspora/diaspora/issues/6289

But yeah I doubt there will be that radical of a design change. I'll check it off for now :+1: It is now really only the `prompts` parameter. I'm working on it!

**Comment by cmrd-senya** [comment]
I have implemented "Authorization Code Flow" support in [my gem](https://github.com/cmrd-senya/diaspora-api).

**Comment by jhass** [comment]
This was merged as part of #8086, thank you!
